src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/DataFormatsTests.cs

```
@@ -0,0 +1,181 @@
+        }
+
+        public static IEnumerable<object[]> GetFormat_KnownString_TestData()
+        {
+            yield return new object[] { DataFormats.Bitmap, "Bitmap", 2 };
+            yield return new object[] { DataFormats.Dib, "DeviceIndependentBitmap", 8 };
+            yield return new object[] { DataFormats.Dif, "DataInterchangeFormat", 5 };
+            yield return new object[] { DataFormats.EnhancedMetafile, "EnhancedMetafile", 14 };
+            yield return new object[] { DataFormats.FileDrop, "FileDrop", 15 };
+            yield return new object[] { DataFormats.Locale, "Locale", 16 };
+            yield return new object[] { DataFormats.MetafilePict, "MetaFilePict", 3 };
+            yield return new object[] { DataFormats.OemText, "OEMText", 7 };
+            yield return new object[] { DataFormats.Palette, "Palette", 9 };
+            yield return new object[] { DataFormats.PenData, "PenData", 10 };
+            yield return new object[] { DataFormats.Riff, "RiffAudio", 11 };
+            yield return new object[] { DataFormats.SymbolicLink, "SymbolicLink", 4 };
+            yield return new object[] { DataFormats.Text, "Text", 1 };
+            yield return new object[] { DataFormats.Tiff, "TaggedImageFileFormat", 6 };
+            yield return new object[] { DataFormats.UnicodeText, "UnicodeText", 13 };
+            yield return new object[] { DataFormats.WaveAudio, "WaveAudio", 12 };
+        }
+
+        [Theory]
+        [MemberData(nameof(GetFormat_KnownString_TestData))]
+        public void DataFormats_GetFormat_InvokeKnownString_ReturnsExpected(string format, string expectedName, int expectedId)
+        {
+            DataFormats.Format result = DataFormats.GetFormat(format);
+            Assert.Same(result, DataFormats.GetFormat(format));
+            Assert.Same(result, DataFormats.GetFormat(format.ToLower()));
+            Assert.Equal(expectedName, result.Name);
+            Assert.Equal(expectedId, result.Id);
```

What would happen if one of these IDs were wrong? Do we have code paths which rely on the ID?

---

Anything that uses `DataFormats.*.Id` in interop code would be incorrect or throw. For this reason we hard code them in `DataFormats.cs`:

```cs
/// <devdoc>
/// Ensures that the Win32 predefined formats are setup in our format list.
/// This is called anytime we need to search the list
/// </devdoc>
/// <internalonly/>
private static void EnsurePredefined()
{
    if (s_formatCount == 0)
    {
        s_formatList = new Format[]
        {
            //         Text name        Win32 format ID      Data stored as a Win32 handle?
            new Format(UnicodeText,  NativeMethods.CF_UNICODETEXT),
            new Format(Text,         NativeMethods.CF_TEXT),
            new Format(Bitmap,       NativeMethods.CF_BITMAP),
            new Format(MetafilePict, NativeMethods.CF_METAFILEPICT),
            new Format(EnhancedMetafile,  NativeMethods.CF_ENHMETAFILE),
            new Format(Dif,          NativeMethods.CF_DIF),
            new Format(Tiff,         NativeMethods.CF_TIFF),
            new Format(OemText,      NativeMethods.CF_OEMTEXT),
            new Format(Dib,          NativeMethods.CF_DIB),
            new Format(Palette,      NativeMethods.CF_PALETTE),
            new Format(PenData,      NativeMethods.CF_PENDATA),
            new Format(Riff,         NativeMethods.CF_RIFF),
            new Format(WaveAudio,    NativeMethods.CF_WAVE),
            new Format(SymbolicLink, NativeMethods.CF_SYLK),
            new Format(FileDrop,     NativeMethods.CF_HDROP),
            new Format(Locale,       NativeMethods.CF_LOCALE)
        };

        s_formatCount = s_formatList.Length;
    }
}
```

---

> would be incorrect or throw.

or? So there could be a silent failure?

---

Well if one Id mapped to a different Id that was registered, we’d try to copy text as an image, for example. If the is was invalid and didn’t exist we would throw when copying